File: src/main/java/net/noyark/scpslserver/jsmod2/math/Vector.java
package net.noyark.scpslserver.jsmod2.math;

import java.util.Objects;

/**
 * @author magiclu550 #(code) jsmod2
 */

public class Vector {

    private double x;

    private double y;

    private double z;

    public static final Vector ZERO = new Vector(0,0,0);

    public static final Vector ONE = new Vector(1,1,1);

    public static final Vector FORWARD = new Vector(0,0,1);

    public static final Vector BACK = new Vector(0,0,-1);

    public static final Vector UP = new Vector(0,1,0);

    public static final Vector DOWN = new Vector(0,-1,0);

    public static final Vector RIGHT = new Vector(1,0,0);

    public static final Vector LEFT = new Vector(-1,0,0);

    public double sqrMagnitude = x*x+y*y+z*z;

    public static double distance(Vector a, Vector b){
        return vectorSub(a,b).getMagnitude();
    }

    public Vector getNormalize(){
        double num = getMagnitude();
        if(num > 9.99999974737875E-06) return vectorFloatDivided(this,num);
        return ZERO;
    }

    public static Vector lerp(Vector a, Vector b, double t){
        t = Math.min(t,1);
        t = Math.max(t,0);

        return lerpUnclamped(a,b,t);
    }

    public static Vector max(Vector a, Vector b){
        return new Vector(
                Math.max(a.x,b.x),
                Math.max(a.y,b.y),
                Math.max(a.z,b.z)
        );
    }

    public static Vector min(Vector a, Vector b){
        return new Vector(
                Math.min(a.x,b.x),
                Math.min(a.y,b.y),
                Math.min(a.z,b.z)
        );
    }

    public static Vector lerpUnclamped(Vector a, Vector b, double t){
        return vectorAdd(a,vectorFloatMulti(vectorSub(a,b),t));
    }

    public static Vector vectorSub(Vector a, Vector b){
        return new Vector(a.x-b.x,a.y-b.y,a.z-b.z);
    }

    public static Vector vectorAdd(Vector a, Vector b){
        return new Vector(a.x+b.x,a.y+b.y,a.z+b.z);
    }

    public static Vector vectorMulti(Vector a, Vector b){
        return new Vector(a.x*b.x,a.y*b.y,a.z*b.z);
    }

    public static Vector vectorDivided(Vector a, Vector b){
        return new Vector(a.x/b.x,a.y/b.y,a.z/b.z);
    }

    public static Vector vectorFloatSub(Vector a, double b){
        return new Vector(a.x-b,a.y-b,a.z-b);
    }

    public static Vector vectorFloatAdd(Vector a, double b){
        return new Vector(a.x+b,a.y+b,a.z+b);
    }

    public static Vector vectorFloatMulti(Vector a, double d){
        return new Vector(a.x*d,a.y*d,a.z*d);
    }


    public static Vector vectorFloatDivided(Vector a, double b){
        return new Vector(a.x/b,a.y/b,a.z/b);
    }

    public static Vector vectorFloatDivided(double a, Vector b){
        return new Vector(a/b.x,a/b.y,a/b.z);
    }

    public static Vector vectorFloatSub(double a, Vector b){
        return new Vector(a-b.x,a-b.y,a-b.z);
    }

    public double getMagnitude(){
        return Math.sqrt(sqrMagnitude);
    }


    public Vector(double x, double y, double z){
        this.x = x;
        this.y = y;
        this.z = z;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector vector = (Vector) o;
        return Double.compare(vector.x, x) == 0 &&
                Double.compare(vector.y, y) == 0 &&
                Double.compare(vector.z, z) == 0 &&
                Double.compare(vector.sqrMagnitude, sqrMagnitude) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y, z, sqrMagnitude);
    }

}

File: src/main/java/net/noyark/scpslserver/jsmod2/math/Vector.scala
package net.noyark.scpslserver.jsmod2.math

/**
  * scala vector
  *
  * @author magiclu550#(code) jsmod2
  */

class Vector (var x:Double,var y:Double,var z:Double){


  val ZERO = new Vector(0, 0, 0)

  val ONE = new Vector(1, 1, 1)

  val FORWARD = new Vector(0, 0, 1)

  val BACK = new Vector(0, 0, -1)

  val UP = new Vector(0, 1, 0)

  val DOWN = new Vector(0, -1, 0)

  val RIGHT = new Vector(1, 0, 0)

  val LEFT = new Vector(-1, 0, 0)

  var sqrMagnitude: Double = x * x + y * y + z * z



  def -(vector: Vector): Vector = new Vector(this.x-vector.x,this.y-vector.y,this.z-vector.z)

  def +(vector: Vector): Vector = new Vector(this.x + vector.x, this.y + vector.y, this.z + vector.z)

  def *(vector: Vector): Vector = new Vector(this.x*vector.x,this.y*vector.y,this.z*this.z)

  def /(vector: Vector): Vector = new Vector(this.x/vector.x,this.y/vector.y,this.z/vector.z)

  def -(d:Double): Vector = new Vector(this.x-d,this.y-d,this.z-d)

  def +(d:Double): Vector = new Vector(this.x+d,this.y+d,this.z+d)

  def *(d:Double): Vector = new Vector(this.x*d,this.y*d,this.z*d)

  def /(d:Double): Vector = new Vector(this.x/d,this.y/d,this.z/d)

  def getNormalize: Vector = {
    val num = getMagnitude
    if (num > 9.99999974737875E-06) return this/num
    ZERO
  }

  def distance(a: Vector, b: Vector): Double = (a-b).getMagnitude

  def lerp(a: Vector, b: Vector, t: Double): Vector = {
    var to = Math.min(t, 1)
    to = Math.max(to, 0)
    lerpUnclamped(a, b, to)
  }

  def lerpUnclamped(a: Vector, b: Vector, t: Double): Vector = a+(a-b)* t

  def getMagnitude: Double = Math.sqrt(sqrMagnitude)

  def min(a: Vector, b: Vector) = new Vector(Math.min(a.x, b.x), Math.min(a.y, b.y), Math.min(a.z, b.z))

  def max(a: Vector, b: Vector) = new Vector(Math.max(a.x, b.x), Math.max(a.y, b.y), Math.max(a.z, b.z))

  def canEqual(other: Any): Boolean = other.isInstanceOf[Vector]

  override def equals(other: Any): Boolean = other match {
    case that: Vector =>
      (that canEqual this) &&
        ZERO == that.ZERO &&
        ONE == that.ONE &&
        FORWARD == that.FORWARD &&
        BACK == that.BACK &&
        UP == that.UP &&
        DOWN == that.DOWN &&
        RIGHT == that.RIGHT &&
        LEFT == that.LEFT
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(ZERO, ONE, FORWARD, BACK, UP, DOWN, RIGHT, LEFT)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
